icu4j/main/core/src/main/java/com/ibm/icu/impl/ICUService.java

```diff
@@ -592,13 +592,15 @@ private Map<String, Factory> getVisibleIDMap() {
                         Factory f = lIter.previous();
                         f.updateVisibleIDs(mutableMap);
                     }
-                    this.idcache = Collections.unmodifiableMap(mutableMap);
+                    Map<String, Factory> result = Collections.unmodifiableMap(mutableMap);
```

Comment:
@gvictor Can we resolve this problem by simply changing `private Map<String, Factory> idcache;` to `private volatile Map<String, Factory> idcache;`?

Reply:
I think not. If I understand the ICUServiceThreadTest correctly, the problem is that `clearCaches()` runs after `this.idcache` is set and before `return  idcache`. Making volatile will likely make the problem more discoverable, but doesn't fix it, as the CPU cache value of `this.idcache` is invalidated by `clearCaches()` during that window. The fix here is to store the newly created cache instance in a local stack reference which can never be null.

Reply:
Thanks for your explanation. It would be nice to add a comment why it makes difference. For meanwhile, I approve this.

Reply:
so your fix is return the "result" in that case instead of this.idcache because idcache could be set to null by other thread before the return but the local result will never be a null regardless how other thread setting idcache, right?

Reply:
Frank, it seems to resolve the flaky tests in our internal CI. If you are satisfied with it, can you help merge it? As I don't have the right to submit this.